packages/workbox-core/_private/DBWrapper.mjs

```diff
@@ -0,0 +1,351 @@
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+import '../_version.mjs';
+
+/**
+ * A class that wraps common IndexedDB functionality in a promise-based API.
+ * It exposes all the underlying power and functionality of IndexedDB, but
+ * wraps the most commonly used features in a way that's much simpler to use.
+ */
+class DBWrapper {
+  /**
+   * @param {string} name
+   * @param {number} version
+   * @param {Object=} [callback]
+   * @param {function(this:DBWrapper, Event)} [callbacks.onupgradeneeded]
+   * @param {function(this:DBWrapper, Event)} [callbacks.onversionchange]
+   *     Defaults to DBWrapper.prototype._onversionchange when not specified.
+   */
+  constructor(name, version, {
+    onupgradeneeded,
+    onversionchange = this._onversionchange,
+  } = {}) {
+    this._name = name;
+    this._version = version;
+    this._onupgradeneeded = onupgradeneeded;
+    this._onversionchange = onversionchange;
+
+    // If this is null, it means the database isn't open.
+    this._db = null;
+  }
+
+  /**
+   * Opens a connected to an IDBDatabase, invokes any onupgradedneeded
+   * callback, and added an onversionchange callback to the database.
+   *
+   * @return {IDBDatabase}
+   */
+  async open() {
+    if (this._db) return;
+
+    this._db = await new Promise((resolve, reject) => {
+      // This flag is flipped to true if the timeout callback runs prior
+      // to the request failing or succeeding. Note: we use a timeout instead
+      // of an onblocked handler since there are cases where onblocked will
+      // never never run. A timeout better handles all possible scenarios:
+      // https://github.com/w3c/IndexedDB/issues/223
+      let openRequestTimedOut = false;
+      setTimeout(() => {
+        openRequestTimedOut = true;
+        reject(new Error('The open request was blocked and timed out'));
+      }, this.OPEN_TIMEOUT);
+
+      const openRequest = indexedDB.open(this._name, this._version);
+      openRequest.onerror = (evt) => reject(openRequest.error);
+      openRequest.onupgradeneeded = (evt) => {
+        if (openRequestTimedOut) {
+          openRequest.transaction.abort();
+          evt.target.result.close();
+        } else if (this._onupgradeneeded) {
+          this._onupgradeneeded(evt);
+        }
+      };
+      openRequest.onsuccess = (evt) => {
+        const db = evt.target.result;
+        if (openRequestTimedOut) {
+          db.close();
+        } else {
+          db.onversionchange = this._onversionchange;
```

Comment:
Not quite. I was confused by this as well, but after looking into it I realized that the purpose of `onversionchange` is so *other* instances can "relinquish" (or optionally not relinquish) their control over the database.

The IDB spec ensures only one version of a database is ever running at a time, so if an `open()` call requests a new version, all existing open references to that database will get the `onversionchange` event (with the assumption that they will call `close()`, so the requester can successfully upgrade). If any of the other open instance don't call `close()` immediately, the requestor gets the `blocked` event (the idea being that it can then show a message to the user asking them to close other open tabs to this site).

The timeout is implemented because there are [cases when an open request will never get any event](https://github.com/w3c/IndexedDB/issues/223), so we have to handle that case too.

The logic in [these tests](https://github.com/GoogleChrome/workbox/blob/194b512aa3b654194ebd344c2cc227f7809d4fda/test/workbox-core/node/_private/test-DBWrapper.mjs#L186-L223) might help make this more clear.